Example-CustomUserGroups-REXX/USRGRP01.rexx

```diff
@@ -0,0 +1,3123 @@
+    )
+
+  /********************************************************************
+   * Traverse the LPARs array to populate LPARsList.
+   * We use the REXX (1-based) idiom  but adjust for the differing
+   * (0-based) idiom of the toolkit.
+   ********************************************************************/
+  Say 'Processing information for ' || LPARs ||,
+      ' LPAR(s)... searching for local'
+
+  Do i = 1 To LPARs
+    nextEntryHandle = GetJSONArrayEntry(LPARArray,i-1)
+
+    LPARlocal = FindJSONValue(,
+      nextEntryHandle,,
+      "request-origin",,
+      HWTJ_BOOLEAN_TYPE,
+    )
+
+    If LPARlocal = 'true' Then Do
+      Say 'Found local LPAR'
+
+      If localLPARfound <> 0 Then Do
+        Say 'Found two local LPARs'
+        Return 0
+      End
+      localLPARfound = TRUE
+
+      lparURI = FindJSONValue(,
+        nextEntryHandle,,
+        "object-uri",,
+        HWTJ_STRING_TYPE,
+      )
+
+      If WasJSONValueFound(lparURI) = FALSE Then Do
+        Say 'Failed to obtain LPAR URI'
+        Return 0
+      End
+
+      lparTargetName = FindJSONValue(,
+        nextEntryHandle,,
+        "target-name",,
+        HWTJ_STRING_TYPE,
+      )
+
+      If WasJSONValueFound(lparTargetName) = FALSE Then Do
+        Say 'Failed to obtain LPAR target name'
+        Return 0
+      End
+
+      LPARname = FindJSONValue(,
+        nextEntryHandle,,
+        "name",,
+        HWTJ_STRING_TYPE,
+      )
+
+      If WasJSONValueFound(LPARname) = FALSE Then Do
+        Say 'Failed to obtain LPAR name'
+        Return 0
+      End
+    End
+  End
+
+  If localLPARfound = FALSE Then Do
+    methodSuccess = FALSE
+    Say 'Failed to retrieve local LPAR info'
+  End
+  Else Do
+    Say
+    Say 'Successfully obtained local LPAR Info:'
+    Say '  name:' || LPARname
+    Say '  uri:' || lparURI
+    Say '  target-name:' || lparTargetName
+    Say
+  End
+
+Return methodSuccess
+
+/**********************************************************************
+ * Function: GetCustomUserGroups
+ *
+ * Purpose: Retrieve the custom user groups visible to this user
+ *
+ * Input: TARGET - The target object to retrieve the custom user groups
+ *                 on
+ *
+ * Output:
+ *      A stem tail variable where the variable's values correspond
+ *      to the following:
+ *
+ *      customUserGroups.0 = N number of custom user groups
+ *      customUserGroups.<Custom Group Idx>.NAME = <Custom_Group_Name1>
+ *      customUserGroups.<Custom Group Idx>.URI = <Custom_Group_URI>
+ *      ...
+ *
+ * Side-Effects: None
+ *
+ **********************************************************************/
+GetCustomUserGroups:
+
+  GROUPS_LIST_KEY = 'groups'
+  GROUP_NAME_KEY = 'name'
+  GROUP_URI_KEY = 'object-uri'
+
+  TARGET = Arg(1)
+
+  customUserGrpResponse = GetRequest(,
+    CUSTOM_USR_GRPS_BASE_URI,,
+    TARGET,
+  )
+
+  If customUserGrpResponse = '' Then Do
+    Say 'Failed to find Custom User Groups '
+    Exit 0

```

Comment:
general comment: you can not just Exit w/out cleaning up the parser storage that was obtain or turning off HWIHOST, that's why Cleanup exist, if you do just exit, you're leaking storage

Reply:
or you can invoke `FatalErrorAndCleanup`

Reply:
why exit 0? the reason i'm digging is because this returns 0 which in z/OS land is success.... similar comment to other spots where you exit 0

Reply:
I agree this probably should not be a 0 but, rather an invocation to fatalError. As for the other spots:

1. Line 329 - If the user simply wished to list the custom user group members then we have nothing more to do so we cleanup and leave the script

2. Line 482 - Exit 0 here is valid as the user wishes to know about the usage of the script and therefore should not throw an error as they did nothing wrong

3. Line 1286 - Exit 0 here is valid since the user has not specified the custom group they wish to target. This can be caused by two reasons. The first is the user failed to specify the custom user group and the second is they do not know the custom user group's name. As a result, of this ambiguity we don't want to throw an error since we can't tell which one it is

4. Line 1405 - Exit 0 here is the right course of action as we are running into a situation where a user wishes to take an action against the custom user group members. Now at this point we have three possible actions they could take:
  1.  They wish to remove a member (if that is the case then fail since that action is not possible)
  2. They wish to add a new member (in this case continue the processing of the script since the add member action occurs later).
  3. They wish to list out the members (in this case since there are no members to list we simply inform users through the say statement at line 1385 and then cleanup and exit the script since there is nothing more we can to do)